source/s3_endpoint.c

```diff
@@ -230,65 +244,72 @@ static struct aws_http_connection_manager *s_s3_endpoint_create_http_connection_
 }
 
 struct aws_s3_endpoint *aws_s3_endpoint_acquire(struct aws_s3_endpoint *endpoint) {
-    AWS_PRECONDITION(endpoint);
+    if (endpoint) {
+        s_s3_endpoint_system_vtable->acquire(endpoint);
+    }
+    return endpoint;
+}
 
-    aws_ref_count_acquire(&endpoint->ref_count);
+static void s_s3_endpoint_acquire(struct aws_s3_endpoint *endpoint) {
+    AWS_PRECONDITION(endpoint);
 
-    return endpoint;
+    aws_s3_client_lock_synced_data(endpoint->client);
+    AWS_ASSERT(endpoint->client_synced_data.ref_count > 0);
+    ++endpoint->client_synced_data.ref_count;
+    aws_s3_client_unlock_synced_data(endpoint->client);
 }
 
 void aws_s3_endpoint_release(struct aws_s3_endpoint *endpoint) {
-    if (endpoint == NULL) {
-        return;
+    if (endpoint) {
+        s_s3_endpoint_system_vtable->release(endpoint);
     }
-
-    aws_ref_count_release(&endpoint->ref_count);
 }
 
-void aws_s3_client_endpoint_release(struct aws_s3_client *client, struct aws_s3_endpoint *endpoint) {
+static void s_s3_endpoint_release(struct aws_s3_endpoint *endpoint) {
     AWS_PRECONDITION(endpoint);
-    AWS_PRECONDITION(client);
-    AWS_PRECONDITION(endpoint->handled_by_client);
+    AWS_PRECONDITION(endpoint->client);
 
     /* BEGIN CRITICAL SECTION */
-    {
-        aws_s3_client_lock_synced_data(client);
-        /* The last refcount to release */
-        if (aws_atomic_load_int(&endpoint->ref_count.ref_count) == 1) {
-            aws_hash_table_remove(&client->synced_data.endpoints, endpoint->host_name, NULL, NULL);
-        }
-        aws_s3_client_unlock_synced_data(client);
+    aws_s3_client_lock_synced_data(endpoint->client);
+
+    AWS_ASSERT(endpoint->client_synced_data.ref_count > 0);
+    bool refcount_is_zero = (--endpoint->client_synced_data.ref_count == 0);
```

Comment:
nit: order of operations seems a bit suspect here. i would prefer to check that its 1 as in old version and only then decrement and remove from hash table. Assert will be compiled away in release builds and without it there is no protection against ref_count wrapping around (but thats very unlikely to happen in real world conditions)